### PR TITLE
lineno fixes

### LINKED ISSUE #922: incorrect line number reported for index out of range error
If you run:
```
function foo()
	x = ones(10)
	print("first access\n")
	x[1]
	print("second access\n")
	x[20]
end
foo()
```
the result is
```
first access
second access
in foo: arrayref: index out of range
 in foo at none:4
```
Note the line number is incorrect. As in #900, this isn't helpful when trying to debug :)

### PR DESCRIPTION
**Pull request by vtjnash**
fixes #922 (wrong line numbers in error expressions)

### CODE REVIEW

File: src/task.c [hunk 1]
```diff
@@ -828,11 +828,6 @@ DLLEXPORT void jl_rethrow_other(jl_value_t *e)
     throw_internal(e);
 }
 
-DLLEXPORT void jl_throw_with_superfluous_argument(jl_value_t *e, int line)
-{
-    jl_throw(e);
-}
-
```
Comment (carnaval) [inline src/task.c]: :-(

File: base/inference.jl [hunk 2]
```diff
@@ -2248,8 +2248,7 @@ function inlineable(f::ANY, e::Expr, atype::ANY, sv::StaticVarInfo, enclosing_as
     end
 
     body = Expr(:block)
-    body.args = filter(x->!((isa(x,Expr) && is(x.head,:line)) || isa(x,LineNumberNode)),
-                       ast.args[3].args::Array{Any,1})
+    body.args = ast.args[3].args::Array{Any,1}
     cost::Int = 1000
```
Comment (ihnorton) [inline base/inference.jl]: Should the cost threshold be adjusted? The top-level expression count is going to increase significantly without the filter.

File: src/codegen.cpp [hunk 3]
```diff
@@ -5923,13 +5908,16 @@ static void init_julia_llvm_env(Module *m)
 
 extern "C" void jl_init_codegen(void)
 {
+    const char *const argv_tailmerge[] = {"", "-enable-tail-merge=0"}; // NOO TOUCHIE; NO TOUCH! See #922
```
Comment (ihnorton) [inline src/codegen.cpp]: Very funny :smile:

**Comment by ihnorton** [comment]
:+1: LGTM

**Comment by Keno** [comment]
Are we sure that disabling tail merging globally is a good idea? What are the performance impacts on other code that benefits from this optimization?

**Comment by vtjnash** [comment]
> Are we sure that disabling tail merging globally is a good idea? What are the performance impacts on other code that benefits from this optimization?

it should result in a very small performance increase (yes, benefit) on some paths in functions that this affects, at the cost of not detecting when the user has duplicated code on multiple function branches (e.g. due to inlining and macros) and folding away the line numbers for that (thus losing on the code-size benefit of this optimization).

we might want to manually merge all of the gc-frame pops into a single BasicBlock to cut down on the number of prologues that get emitted. (e.g. [this pass](https://github.com/llvm-mirror/llvm/blob/master/lib/Transforms/Utils/UnifyFunctionExitNodes.cpp), but with more specificity for Julia)

File: base/inference.jl (hunk 2 above)
Comment (vtjnash) [inline base/inference.jl]: possibly, it somewhat affects the expected average number of symbols per line (https://github.com/JuliaLang/julia/blob/d3af973d439e8f3628b63c7e305fdf5edf165ae2/base/inference.jl#L2624), but it's probably not too significant. the inliner doesn't want to inline long functions, and each `:line` Expr is pretty small. I think I'll wait to see if it is a concern.

**Comment by vtjnash** [comment]
also include https://github.com/JuliaLang/julia/commit/a70d7cfafe049588a3d9d9b77efc331e629aabdd when backporting

**Comment by tkelman** [comment]
removing the backport label given #14186

**Pull request merged**